lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -388,201 +378,529 @@ export class UserAgentApplication {
-              this.promptUser(urlNavigate);
-          });
+    );
   }
 
+  //#endregion
+
+  //#region Popup Flow
+
   /**
    * Initiate the login process by opening a popup window.
    * @param {Array.<string>} scopes - Permissions you want included in the access token. Not all scopes are  guaranteed to be included in the access token returned.
    * @param {string} extraQueryParameters - Key-value pairs to pass to the STS during the interactive authentication flow.
    * @returns {Promise.<string>} - A Promise that is fulfilled when this function has completed, or rejected if an error was raised. Returns the token or error.
    */
+  // TODO: params to accept AuthRequest object instead
   loginPopup(scopes ?: Array<string>, extraQueryParameters?: string): Promise<string> {
     /*
     1. Create navigate url
     2. saves value in cache
     3. redirect user to AAD
      */
     return new Promise<string>((resolve, reject) => {
+      // Fail if login is already in progress
       if (this._loginInProgress) {
+        // TODO: Return custom error object here in future
         reject(ErrorCodes.loginProgressError + Constants.resourceDelimeter + ErrorDescription.loginProgressError);
         return;
       }
 
+      // Validate scopes
+      // TODO: Replace with new validation pattern
       if (scopes) {
         const isValidScope = this.validateInputScope(scopes);
         if (isValidScope && !Utils.isEmpty(isValidScope)) {
+          // TODO: Return custom error object here in future
           reject(ErrorCodes.inputScopesError + Constants.resourceDelimeter + ErrorDescription.inputScopesError);
           return;
         }
 
         scopes = this.filterScopes(scopes);
       }
 
-        var idTokenObject;
-        idTokenObject = this.extractADALIdToken();
-        if (idTokenObject && !scopes) {
-            this._logger.info("ADAL's idToken exists. Extracting login information from ADAL's idToken ");
-            extraQueryParameters = Utils.constructUnifiedCacheExtraQueryParameter(idTokenObject, extraQueryParameters);
-            this._silentLogin = true;
-            this.acquireTokenSilent([this.clientId], this.authority, this.getUser(), extraQueryParameters)
-                .then((idToken) => {
-                    this._silentLogin = false;
-                    this._logger.info("Unified cache call is successful");
-                    resolve(idToken);
-                }, (error) => {
-                    this._silentLogin = false;
-                    this._logger.error("Error occurred during unified cache ATS");
-                    this.loginPopupHelper(resolve, reject, scopes, extraQueryParameters);
-                });
-        }
-         else {
-            this.loginPopupHelper(resolve, reject, scopes, extraQueryParameters );
-        }
-      });
-  }
+      // Extract ADAL id_token if it exists
+      var idTokenObject;
```

Comment:
this is done elsewhere duplicated

Reply:
or close to it below

Reply:
There is a lot of duplicated code between redirect and popup. We will try to remove as much of this as possible in future PRs.